app/waf/src/controllers/adc.controller.ts

```diff
@@ -334,47 +335,43 @@ export class AdcController extends BaseController {
       userToken: await this.reqCxt.get(WafBindingKeys.Request.KeyUserToken),
       tenantId: await this.reqCxt.get(WafBindingKeys.Request.KeyTenantId),
     };
+    this.adcStCtr = new AdcStateCtrlr(adc, addonReq);
 
     switch (Object.keys(actionBody)[0]) {
       case 'create':
-        if (adc.status !== 'NONE' && adc.status !== 'POWERERR')
-          throw new HttpErrors.BadRequest(
-            `Adc status is ' ${
-              adc.status
-            }. Cannot repeat 'create' on the same ADC.`,
-          );
-
-        this.createOn(adc, addonReq);
-        return {id: adc.id};
+        if (await this.adcStCtr.readyTo(AdcState.POWERON)) {
+          this.createOn(adc, addonReq);
+          return {id: adc.id};
+        }
 
       case 'delete':
         this.deleteOn(adc, addonReq);
         return {id: adc.id};
 
       case 'setup':
-        this.setupOn(adc, addonReq);
-        return {id: adc.id};
+        if (await this.adcStCtr.readyTo(AdcState.ONBOARDED)) {
+          this.setupOn(adc, addonReq);
+          return {id: adc.id};
+        }
 
       default:
         throw new HttpErrors.BadRequest(
           'Not supported: ' + Object.keys(actionBody)[0],
         );
     }
+
+    let msg = `Cannot do '${Object.keys(actionBody)[0]}' on adc ${
+      adc.id
+    }, state: ${adc.status}, lasterr: ${adc.lastErr}`;
+    this.logger.error(msg);
+    throw new HttpErrors.UnprocessableEntity(msg);
   }
 
   private async setupOn(adc: Adc, addon: AddonReqValues): Promise<void> {
-    let bigipMgr = await BigIpManager.instanlize({
-      username: adc.management!.username,
-      password: adc.management!.password,
-      ipAddr: adc.management!.ipAddress,
-      port: adc.management!.tcpPort,
-    });
-    let bigipReady = async (): Promise<boolean> => {
-      return await bigipMgr.getSys().then(() => {
-        return true;
-      });
-    };
-    await checkAndWait(bigipReady, 240).then(
+    await checkAndWait(
+      () => this.adcStCtr.readyTo(AdcState.ONBOARDED).then(() => true),
```

Comment:
is this line and line 352 duplicated.

Reply:
Right. Done for that. Thanks.
The usage of StateController is not made fully. There are 2 places which are using it. Later will apply all over the code.

Reply:
👌